core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -245,161 +433,159 @@ else if (state == CANCELED) {
+                    // explain analyze stage is finished
+                    if (finishedStages.containsAll(remotelyScheduledStages)) {
+                        stages.values().stream()
+                                .filter(stage -> finishedStages.contains(stage.getStageId()))
+                                .forEach(SqlStageExecution::transitionToFinished);
+                    }
+                }
             });
         }
 
-        // when query is done or any time a stage completes, attempt to transition query to "final query info ready"
-        queryStateMachine.addStateChangeListener(newState -> {
-            if (newState.isDone()) {
-                queryStateMachine.updateQueryInfo(Optional.ofNullable(getStageInfo()));
-            }
-        });
-        for (SqlStageExecution stage : stages.values()) {
-            stage.addFinalStageInfoListener(status -> queryStateMachine.updateQueryInfo(Optional.ofNullable(getStageInfo())));
-        }
-    }
-
-    private static void updateQueryOutputLocations(QueryStateMachine queryStateMachine, OutputBufferId rootBufferId, Set<RemoteTask> tasks, boolean noMoreExchangeLocations)
-    {
-        Map<TaskId, URI> bufferLocations = tasks.stream()
-                .collect(toImmutableMap(RemoteTask::getTaskId, task -> uriBuilderFrom(task.getTaskStatus().getSelf()).appendPath("results").appendPath(rootBufferId.toString()).build()));
-        queryStateMachine.updateOutputLocations(bufferLocations, noMoreExchangeLocations);
+        return new PipelinedScheduler(
+                queryStateMachine,
+                executionSchedule,
+                schedulersBuilder.build(),
+                schedulerStats,
+                executions);
     }
 
-    private List<SqlStageExecution> createStages(
-            ExchangeLocationsConsumer parent,
-            AtomicInteger nextStageId,
+    private PipelinedStageExecution createPipelinedExecution(
             StageExecutionPlan plan,
-            NodeScheduler nodeScheduler,
-            RemoteTaskFactory remoteTaskFactory,
-            Session session,
-            int splitBatchSize,
+            Map<PlanFragmentId, OutputBufferManager> outputBufferManagers,
+            Map<PlanFragmentId, Optional<int[]>> bucketToPartitionMap,
             Function<PartitioningHandle, NodePartitionMap> partitioningCache,
-            NodePartitioningManager nodePartitioningManager,
-            ExecutorService queryExecutor,
-            ScheduledExecutorService schedulerExecutor,
-            FailureDetector failureDetector,
-            NodeTaskMap nodeTaskMap,
-            ImmutableMap.Builder<StageId, StageScheduler> stageSchedulers,
-            ImmutableMap.Builder<StageId, StageLinkage> stageLinkages)
+            TaskLifecycleListener taskLifecycleListener,
+            ImmutableList.Builder<PipelinedStageExecution> executions,
+            ImmutableMap.Builder<StageId, StageScheduler> schedulers)
```

Comment:
stageSchedulers?

Reply:
Generally I think it is worthy to use `stage` in variable names for `StageSchedulers` and `*StageExectionits`
to differentiate from `*QueryScheduler` and `*QueryExecution`

Reply:
> stageSchedulers?

Changed in the next commit